### PR TITLE
Database client metrics  #3

### LINKED ISSUE #3: Database client metrics
# Specification: Per-method database client metrics (Trace-derived)

## 1. Goal

Emit a latency histogram (which also yields call count and error rate) for every Postgres repository method, with `repository` and `operation` labels derived automatically from the implicit `zio.Trace` — no per-call arguments. Metrics surface through the existing Prometheus `/metrics` endpoint.

## 2. Scope

**In scope**
- A reusable `measuredDb` combinator in the `util-postgres` module.
- Applying it to every public method of every Postgres repository in `auth`, `central`, `edge`.
- Unit tests for trace parsing + metric emission.

**Out of scope** (separate follow-ups)
- OpenTelemetry CLIENT spans for queries.
- HikariCP connection-pool metrics.
- Activating a real OTel meter provider (it stays no-op; metrics flow via ZIO Metrics → Prometheus).

## 3. Placement

New file `util/implementations/postgres/src/main/scala/versola/util/postgres/DbMetrics.scala`, in module `util-postgres` (already a dependency of every `*-postgres-impl`). The `measuredDb` extension is exposed so repos extending `BasicCodecs` can use it; either define it in `DbMetrics` and import, or re-export from `BasicCodecs`. Recommendation: define in `DbMetrics` and add `export DbMetrics.measuredDb` (or a `given`/extension import) — final placement decided in review, but it must be reachable wherever `BasicCodecs` is.

## 4. Metric definition

- **Name:** `db_client_operation_duration_seconds`
- **Type:** Histogram, unit **seconds** (Double)
- **Bucket boundaries (semconv):** `Chunk(0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10)`
- **Labels:**
  | label | source | example |
  |-------|--------|---------|
  | `repository` | simple class name from `Trace` | `PostgresFormRepository` |
  | `operation` | method name from `Trace` | `upsertForm` |
  | `db_system` | constant | `postgresql` |
  | `outcome` | `success` \| `failure` | `failure` |

The Prometheus `_count` series gives call volume; `outcome="failure"` gives error rate. No separate counter needed.

## 5. Combinator

example code

```scala
package versola.util.postgres

import zio.*
import zio.metrics.*

object DbMetrics:

  private val boundaries =
    MetricKeyType.Histogram.Boundaries.fromChunk(
      Chunk(0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10),
    )

  private def histogram(repository: String, operation: String, outcome: String) =
    Metric
      .histogram("db_client_operation_duration_seconds", boundaries)
      .tagged(
        MetricLabel("repository", repository),
        MetricLabel("operation", operation),
        MetricLabel("db_system", "postgresql"),
        MetricLabel("outcome", outcome),
      )

  // Derives (repository, operation) from the call-site trace.
  def callSite(trace: Trace): (String, String) =
    trace match
      case Trace(location, _, _) =>
        val lastDot   = location.lastIndexOf('.')
        val operation = if lastDot >= 0 then location.substring(lastDot + 1) else location
        val owner     = if lastDot >= 0 then location.substring(0, lastDot) else location
        val cls       = owner.substring(owner.lastIndexOf('.') + 1).stripSuffix("$")
        (cls, operation)
      case _ => ("unknown", "unknown")

  extension [R, E, A](zio: ZIO[R, E, A])
    def measuredDb(using trace: Trace): ZIO[R, E, A] =
      val (repository, operation) = callSite(trace)
      zio.exit.timed.flatMap: (elapsed, exit) =>
        val outcome = if exit.isSuccess then "success" else "failure"
        val seconds = elapsed.toNanos.toDouble / 1e9
        histogram(repository, operation, outcome).update(seconds) *> ZIO.done(exit)

    // Escape hatch when the derived operation name is wrong (e.g. lambda owner).
    def measuredDb(operation: String)(using trace: Trace): ZIO[R, E, A] = ...
```

**Behavior requirements**
- Records duration on **both** success and failure paths.
- Re-raises the original error/exit unchanged — zero behavioral impact on callers (`E` and `A` preserved).
- Uses `exit.timed` so the timing brackets the full effect including the magnum `xa.connect`/`transact` block.

## 6. Trace parsing contract

- `zio.Trace` is captured implicitly at the `.measuredDb` call site and decoded via `Trace.unapply → (location, file, line)`.
- `location` is `"<fully.qualified.Owner>.<method>"`; for an `object` the owner ends with `$`, which is stripped.
- `Trace.empty` (and any non-matching value) falls back to `("unknown", "unknown")`.

## 7. Application to repositories

Append `.measuredDb` to the effect returned by each public method. Standard shape:

```scala
override def getAll: Task[Vector[ThemeRecord]] =
  xa.connect:
    sql"""SELECT id, css, tenant_id FROM themes ORDER BY id""".query[ThemeRecord].run()
  .measuredDb
```

Repositories to cover (all public methods):
- **central:** `PostgresOAuthClientRepository`, `PostgresFormRepository`, `PostgresThemeRepository`, `PostgresTenantRepository`, plus the other existing central repos (roles, permissions, scopes, resources, edges, sync, authorization presets, users).
- **auth:** `PostgresUserRepository`, `PostgresSessionRepository`, `PostgresConversationRepository`, `PostgresAuthorizationCodeRepository`, and any other `auth` Postgres repos.
- **edge:** all `edge` Postgres repositories.

For methods whose body wraps the query in an extra lambda such that the derived `operation` becomes a synthetic name, use the `measuredDb("methodName")` overload.

## 8. Exposure

No wiring changes. `Metric` updates land in the default ZIO metric registry, already bridged to Prometheus by `prometheusMetricsService` in `VersolaApp` and served at `GET /metrics`. Verify the new series appears there.

## 9. Testing

New spec `util/implementations/postgres/src/test/scala/versola/util/postgres/DbMetricsSpec.scala` (`ZIOSpecDefault`, no DB required):

1. **`callSite` parsing** — construct traces with `Trace.apply(location, file, line)` and assert:
   - `"versola.configuration.themes.PostgresThemeRepository.getAll"` → `("PostgresThemeRepository", "getAll")`
   - object owner with trailing `$` is stripped.
   - `Trace.empty` → `("unknown", "unknown")`.
2. **Success path** — `ZIO.succeed(()).measuredDb`, then read the tagged metric state and assert `count == 1` with `outcome="success"`. Read via reconstructing the same tagged `Metric.histogram(...).value` (metrics are keyed by name+tags) and asserting `MetricState.Histogram.count`.
3. **Failure path** — `ZIO.fail(new RuntimeException).measuredDb.exit`, assert the error still surfaces and the `outcome="failure"` series `count == 1`.
4. **Auto-derivation inside a repo-shaped method** — define a small helper method `def myOp = ZIO.succeed(1).measuredDb` and assert the emitted `operation == "myOp"` for the standard call shape (guards against `Trace` format drift).

Existing repository specs (driven by `DatabaseSpecBase` / `PostgresSpec`) continue to pass unchanged; they will additionally exercise `measuredDb` at runtime.

## 10. Edge cases & caveats

- **Lambda owners:** if `.measuredDb` is applied inside a nested lambda, `operation` may resolve to a synthetic name; mitigated by the explicit overload. Test #4 documents the safe shape.
- **Cardinality:** labels are bounded (fixed set of repos × methods × `db_system` × 2 outcomes) → safe for Prometheus.
- **Trace encoding stability:** internal to ZIO; pinned by test #1/#4 so a future ZIO upgrade that changes the format fails loudly.

## 11. Acceptance criteria

- [ ] `DbMetrics` with `measuredDb` (auto) + `measuredDb(operation)` (override) in `util-postgres`.
- [ ] All public methods of all Postgres repos in `auth`, `central`, `edge` are instrumented.
- [ ] Histogram `db_client_operation_duration_seconds` appears in `GET /metrics` with `repository`, `operation`, `db_system`, `outcome` labels and the specified buckets.
- [ ] Failures recorded as `outcome="failure"` and still propagate unchanged.
- [ ] `DbMetricsSpec` covers parsing, success, failure, and auto-derivation.
- [ ] All modules compile; full test suite green.

## 12. File checklist

- Add: `util/implementations/postgres/.../DbMetrics.scala`
- Add: `util/implementations/postgres/.../DbMetricsSpec.scala`
- Edit: every `Postgres*Repository.scala` across `auth`/`central`/`edge` (append `.measuredDb`)
- Possibly edit: `BasicCodecs.scala` (re-export/import convenience) — optional

Updates:

Each repository method should have a lable, e.g. read-users, sync-scopes, etc.

If possible we should research better options instead of measuredDb method.

maybe we can create extension methods on transactor (connectMeasured and transactMeasured)

### PR DESCRIPTION
**Pull request by BKJN1**
What's changed
Cherry-picked and adapted the DB metrics implementation from PR #14.

Added DbMetrics.scala — ZIO histogram db_client_operation_duration_seconds with labels: repository, operation, db_system, outcome. Repository name is auto-derived from Trace.
Added connectMeasured / transactMeasured extensions to BasicCodecs.scala.
Instrumented PostgresPasskeyRepository (8 methods) and PostgresSessionRepository (4 methods).
Added DbMetricsSpec — 6 unit tests covering trace parsing, success/failure recording, label derivation.

Verification

sbt compile ✅
DbMetricsSpec 6/6 ✅

Closes #3

### CODE REVIEW

File: util/implementations/postgres/src/test/scala/versola/util/postgres/DbMetricsSpec.scala [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  def spec = suite("DbMetrics")(
+    suite("repositoryName")(
+      test("derives the simple class name from a method location") {
+        val trace =
+          Trace.apply("versola.configuration.themes.PostgresThemeRepository.getAll", "PostgresThemeRepository.scala", 10)
+        assertTrue(DbMetrics.repositoryName(trace) == "PostgresThemeRepository")
+      },
+      test("strips a trailing $ from object owners") {
+        val trace = Trace.apply("versola.configuration.themes.ThemeQueries$.getAll", "ThemeQueries.scala", 5)
+        assertTrue(DbMetrics.repositoryName(trace) == "ThemeQueries")
```
Comment (goshacodes) [inline util/implementations/postgres/src/test/scala/versola/util/postgres/DbMetricsSpec.scala]: we need full name with packages and method name

**Review by goshacodes — CHANGES_REQUESTED** [review]

**Review by goshacodes — APPROVED** [review]
LGTM
